server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1131,72 +1441,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+                                                    final ShardSnapshotStatus existing = entry.shards().get(key.value);
+                                                    if (existing.equals(ShardSnapshotStatus.UNASSIGNED_WAITING)
+                                                            && reAssignedShardIds.add(key.value)) {
+                                                        updatedAssignmentsBuilder.put(key.value, shardAssignments.get(key.value));
+                                                    } else {
+                                                        updatedAssignmentsBuilder.put(key.value, existing);
+                                                    }
+                                                }
+                                                snapshotEntries.add(entry.withShards(updatedAssignmentsBuilder.build()));
+                                                updatedQueuedSnapshot = true;
+                                                break;
+                                            }
+                                        }
+                                        if (updatedQueuedSnapshot == false) {
+                                            // Nothing to update in this snapshot so we just add it as is
+                                            snapshotEntries.add(entry);
+                                        }
+                                    } else {
+                                        // Entry is already completed so we will finalize it now that the delete doesn't block us after
+                                        // this CS update finishes
+                                        newFinalizations.add(entry);
+                                        snapshotEntries.add(entry);
+                                    }
+                                } else {
+                                    // Entry is for another repository we just keep it as is
+                                    snapshotEntries.add(entry);
+                                }
+                            }
+                            updatedSnapshotsInProgress = new SnapshotsInProgress(snapshotEntries);
+                        }
+                        return ClusterState.builder(currentState).putCustom(SnapshotDeletionsInProgress.TYPE, deletions)
+                                .putCustom(SnapshotsInProgress.TYPE, updatedSnapshotsInProgress).build();
                     }
                 }
                 return currentState;
             }
 
             @Override
             public void onFailure(String source, Exception e) {
-                logger.warn(() -> new ParameterizedMessage("{} failed to remove snapshot deletion metadata", snapshotIds), e);
-                if (listener != null) {
-                    listener.onFailure(e);
+                logger.warn(() -> new ParameterizedMessage("{} failed to remove snapshot deletion metadata", deleteEntry), e);
```

Comment:
This is a scenario that was complicated before but became very complicated now, both fail-over on delete and create. We could put more effort into this by doing things like https://github.com/elastic/elasticsearch/pull/54350 but I think it's not a priority relative to getting concurrent operations work functionally so I just went with the brute-force fail everything on master fail-over approach here as that's the current behavior for single snapshots.